src/Compilers/VisualBasic/Portable/Symbols/ConstraintsHelper.vb

```diff
@@ -610,6 +649,140 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
+                    ' We can cut short if the original definition's base doesn't contain type parameters.
+                    Dim currentType = type
+                    Dim baseType = currentType.BaseTypeNoUseSiteDiagnostics
+                    Dim result = True
+
+                    While baseType IsNot Nothing AndAlso IsOrRefersToTypeParameter(currentType.OriginalDefinition.BaseTypeNoUseSiteDiagnostics)
+                        result = result And CheckTupleNamesConstraints(baseType, diagnosticsBuilder, TupleNamesCheckKind.Deep, skipList)
+                        currentType = baseType
+                        baseType = currentType.BaseTypeNoUseSiteDiagnostics
+                    End While
+                    Return result
+
+                Case TupleNamesCheckKind.DeepExceptTopLevel
+                    skipList.Add(type)
+                    If Not CheckTupleNamesConstraintsCore(type, diagnosticsBuilder) Then
+                        Return False
+                    End If
+
+                    Dim baseType = type.BaseTypeNoUseSiteDiagnostics
+                    If baseType Is Nothing OrElse Not IsOrRefersToTypeParameter(type.OriginalDefinition.BaseTypeNoUseSiteDiagnostics) Then
+                        Return True
+                    End If
+                    Return CheckTupleNamesConstraints(baseType, diagnosticsBuilder, TupleNamesCheckKind.Deep, skipList)
+
+                Case TupleNamesCheckKind.Deep
+                    Dim failedType = type.VisitType(
+                        Function(t As TypeSymbol, arg As (diag As ArrayBuilder(Of TypeParameterDiagnosticInfo), skip As HashSet(Of TypeSymbol)))
+                            Return Not CheckTupleNamesConstraints(t, arg.diag, TupleNamesCheckKind.Shallow, arg.skip)
+                        End Function, (diag:=diagnosticsBuilder, skip:=skipList))
+                    Return failedType Is Nothing
+
+        Case Else
+        Throw ExceptionUtilities.UnexpectedValue(tupleNamesCheckKind)
+        End Select
+        End Function
+
+        ''' <summary>
+        ''' Checks a single type for tuple name constraints, without recursing into its base types or constituent types.
+        ''' </summary>
+        Private Function CheckTupleNamesConstraintsCore(type As TypeSymbol,
+                                                    <[In], Out> ByRef diagnosticsBuilder As ArrayBuilder(Of TypeParameterDiagnosticInfo)) As Boolean
+
+            Dim baseType As NamedTypeSymbol = type.BaseTypeNoUseSiteDiagnostics
+            If baseType Is Nothing Then
+                Return True
+            End If
+
+            Dim interfaces = type.InterfacesAndTheirBaseInterfacesNoUseSiteDiagnostics
+            If interfaces.IsEmpty Then
+                Return True
+            End If
+
+            Dim baseInterfaces = New Dictionary(Of NamedTypeSymbol, NamedTypeSymbol)(EqualsIgnoringComparer.InstanceIgnoringTupleNames)
```

Comment:
>Dim baseInterfaces = New Dictionary(Of NamedTypeSymbol, NamedTypeSymbol)(EqualsIgnoringComparer.InstanceIgnoringTupleNames) [](start = 12, length = 123)

Please make a change similar to C#, i.e. return early if baseType.AllInterfacesNoUseSiteDiagnostics is empty.